drivers/serial/uart_smartbond.c

```diff
@@ -0,0 +1,501 @@
+	{   38400, 0x00003401 },
+	{   28800, 0x00004507 },
+	{   19200, 0x00006803 },
+	{   14400, 0x00008a0e },
+	{    9600, 0x0000d005 },
+	{    4800, 0x0001a00b },
+};
+
+struct uart_smartbond_cfg {
+	UART2_Type *regs;
+	int periph_clock_config;
+	const struct pinctrl_dev_config *pcfg;
+	bool hw_flow_control_supported;
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	void (*irq_config_func)(const struct device *dev);
+#endif
+};
+
+struct uart_smartbond_data {
+	struct uart_config current_config;
+	struct k_spinlock lock;
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	uart_irq_callback_user_data_t callback;
+	void *cb_data;
+	uint32_t flags;
+	uint8_t rx_enabled;
+	uint8_t tx_enabled;
+#endif
+};
+
+static int uart_smartbond_poll_in(const struct device *dev, unsigned char *p_char)
+{
+	const struct uart_smartbond_cfg *config = dev->config;
+	struct uart_smartbond_data *data = dev->data;
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	if ((config->regs->UART2_USR_REG & UART2_UART2_USR_REG_UART_RFNE_Msk) == 0) {
+		k_spin_unlock(&data->lock, key);
+		return -1;
+	}
+
+	*p_char = config->regs->UART2_RBR_THR_DLL_REG;
+
+	k_spin_unlock(&data->lock, key);
+
+	return 0;
+}
+
+static void uart_smartbond_poll_out(const struct device *dev, unsigned char out_char)
+{
+	const struct uart_smartbond_cfg *config = dev->config;
+	struct uart_smartbond_data *data = dev->data;
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	while (!(config->regs->UART2_USR_REG & UART2_UART2_USR_REG_UART_TFNF_Msk)) {
+		/* Wait until FIFO has free space */
+	}
+
+	config->regs->UART2_RBR_THR_DLL_REG = out_char;
+
+	k_spin_unlock(&data->lock, key);
+}
+
+static int uart_smartbond_configure(const struct device *dev,
+				  const struct uart_config *cfg)
+{
+	const struct uart_smartbond_cfg *config = dev->config;
+	struct uart_smartbond_data *data = dev->data;
+	uint32_t baudrate_cfg = 0;
+	k_spinlock_key_t key;
+	uint32_t reg_val;
+	int err;
+	int i;
+
+	if ((cfg->parity != UART_CFG_PARITY_NONE && cfg->parity != UART_CFG_PARITY_ODD &&
+	     cfg->parity != UART_CFG_PARITY_EVEN) ||
+	    (cfg->stop_bits != UART_CFG_STOP_BITS_1 && cfg->stop_bits != UART_CFG_STOP_BITS_2) ||
+	    (cfg->data_bits != UART_CFG_DATA_BITS_5 && cfg->data_bits != UART_CFG_DATA_BITS_6 &&
+	     cfg->data_bits != UART_CFG_DATA_BITS_7 && cfg->data_bits != UART_CFG_DATA_BITS_8) ||
+	    (cfg->flow_ctrl != UART_CFG_FLOW_CTRL_NONE &&
+	     cfg->flow_ctrl != UART_CFG_FLOW_CTRL_RTS_CTS)) {
+		return -ENOTSUP;
+	}
+
+	/* Flow control is not supported on UART */
+	if (cfg->flow_ctrl == UART_CFG_FLOW_CTRL_RTS_CTS &&
+	    !config->hw_flow_control_supported) {
+		return -ENOTSUP;
+	}
+
+	/* Lookup configuration for baudrate */
+	for (i = 0; i < ARRAY_SIZE(uart_smartbond_baudrate_table); i++) {
+		if (uart_smartbond_baudrate_table[i].baudrate == cfg->baudrate) {
+			baudrate_cfg = uart_smartbond_baudrate_table[i].cfg;
+			break;
+		}
+	}
+
+	if (baudrate_cfg == 0) {
+		return -ENOTSUP;
+	}
+
+	key = k_spin_lock(&data->lock);
```

Comment:
question: why is spinlock required here?

Reply:
registers that are used for configuration have different function during normal usage, thus they are protected here and in other places that access them (e.g. FIFO access)